[Add] FirebaseAPIManager

DalioKim opened the pull request:
# 주요 변경사항
	- FirebaseFirestore 라이브러리를 이용한 제네릭 네트워크 계층 FirebaseAPIManager 구현
	- Firestore+Combine, CollectionReference+AddQueries 등 네트워크 계층 유틸파일
	- 파이어베이스에 Request요청시 사용할 프로토콜 구현

# 중점적으로 봐줬으면 하는 부분
	- 유틸파일간 종속성 트리 
	- 폴더 및 파일 구분 적합성
	- 네이밍 적합성
	- 예외처리해야할 에러
	- Post 요청 후 리턴받는 퍼블리셔 타입

# 추가 작업 예정사항
	- Network Error 정의
	- TDD
	- FirebaseAPIManager + Auth 및 API 콘테이너 추가
	- dateFormatter 추가

# 사용 예시
        1. 먼저 다음과 같이 파이어베이스 콘솔에 테이블을 추가합니다. 

<img width="897" alt="스크린샷 2022-11-05 오전 12 00 11" src="https://user-images.githubusercontent.com/56388189/200007108-7c772f8f-7af9-48b8-b46b-c519c1e10ebf.png">

(2개 이상의 쿼리 요청시 정렬을 위한 색인 설정이 필요합니다.)

<img width="895" alt="스크린샷 2022-11-08 오후 5 34 48" src="https://user-images.githubusercontent.com/56388189/200515188-69d7d61f-e43c-49fe-85e1-0c24817e47b3.png">

        2. 테스트 편의를 위해 APIs 파일을 생성합니다.(뷰모델에 구현하셔도 됩니다.)
        
```
import Combine

class APIs {
    static var cancelBag = Set<AnyCancellable>()

    static func testFetchRunningRecord(request: RunningRecordRequest) -> AnyPublisher<[RunningRecordResponse], Error> {
        FirebaseAPIManager.shared.fetch(request: request, as: RunningRecordResponse.self)
    }
    
    static func testPostRunningRecord(request: RunningRecordRequest) -> AnyPublisher<Bool, Error> {
        FirebaseAPIManager.shared.post(request: request)
    }
}
```
       3.테이블의 request와 response 파일을 생성합니다.

```
struct RunningRecordRequest {
    let uid: String
    let date: [String]
}

extension RunningRecordRequest: FireStoreRequest {
    var collectionName: String {
        "RunningRecord"
    }
}

import Foundation

struct RunningRecordResponse: Decodable {
    let uid: String?
    let distance: Int?

    enum Keys: String, CodingKey {
        case uid
        case distance
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        uid = try container.decodeIfPresent(String.self, forKey: .uid) ?? ""
        distance = try container.decodeIfPresent(Int.self, forKey: .distance) ?? 0
    }
}

```

       4.테스트할 화면에  onAppear 연산자를 이용해 테스트를 구현합니다.
       (편의를 위해서입니다. 뷰모델로 제대로 구현시에는 다르게 하셔야 합니다.)

```
import SwiftUI

// MARK: - JustStartView

struct JustStartView: View {
    @State private var isPresentedRunGuideDetailDescriptionView = false

    var body: some View {
        ZStack {
            MapView().hide(isPresentedRunGuideDetailDescriptionView)
            VStack {
                RunGuideTabView(isPresentedRunGuideDetailDescriptionView: $isPresentedRunGuideDetailDescriptionView).padding(.top, 5)
                Spacer()
                BottomButtonView().padding(.bottom, 30)
                    .onAppear {
                        APIs.testFetchRunningRecord(request: RunningRecordRequest(uid: "test", date: ["2022/10/31"]))
                            .sink { error in
                                print("testFetchRunningRecord error \(error)")
                            } receiveValue: { result in
                                print("testFetchRunningRecord result \(result)")
                            }
                            .store(in: &APIs.cancelBag)
                    }
            }
        }
    }
}

```

DalioKim committed:
FEAT: APIManger & APIManger+AuthRequest

1. APIManger container 클래스 선언
2. APIManger의 extension과 enum case를 통해 generic 구성
3. NetworkError 추가
4. APIManger+AuthRequest 추가

DalioKim committed:
FIX: APIManger 타입 수정

DalioKim committed:
REFACTOR: FirebaseAPIManager 분리

DalioKim committed:
REFACTOR:
1.requestAuth guard let 바인딩등 로직 변경
2. FirebaseAPIManager+Auth extension으로 변경
FEAT:
1. FirebaseAPIManager+GetDocuments 추가
2. CollectionReference+AddQueries, JSONDecoder+Decode, Query+Combine 등 추가
3. OptionNameSpaces 추가

DalioKim committed:
REFACTOR:
1.getDocuments의 as 파라미터 타입 수정
2. DATA/Helper 루트 추가 및 Utils,Helper 경로의 파일 구성 변경

FEAT:
1. getDocument 추가
2. getDocument에 필요한 Query+map 메서드 추가

DalioKim committed:
FIX: getDocument 리턴 타입 수정

DalioKim committed:
REFACTOR:
1. 네이밍 수정

FEAT:
1. post 및 addDocumentWithAnyPublisher 메서드 구현
2 Encodable+Dictionary 구현

DalioKim committed:
REFACTOR:
1. 임시코드 삭제
2. QueryOption, CollectionReference+AddQueries에 범위 관련(기간등) case 추가

DalioKim committed:
REFACTOR: FireStoreRequest 구현 및 적용
1.Encodable+FireStore의 queries와 FireStoreRequest을 이용하여, 기존 FirebaseAPIManager 로직 대체
2. getDocumentWithAnyPublisher 메서드 CollectionReference로 변경
3. 그외 네이밍, 파라미터, 바인딩 로직 수정

DalioKim committed:
REFACTOR: 강제해제 제거 및 case 구분

DalioKim committed:
REFACTOR: FirebaseAPIManager 및 유틸 리팩토링
1. 메서드, 파라미터 네이밍 수정(fetch -> fetchOnce, getDocumentWithPublisher의 type -> as 등)
2. Encodable+queries 파일명 수정 및 내부 switch문 로직 수정
3. 불필요한 코드 삭제 및 컨벤션 수정

DalioKim committed:
REFACTOR: Firestore 로직 변경

1. Firestore+Combine 추가(네트워크 요청 및 디코드 로직 구분)
2. CollectionReference+Combine 삭제
3. FireStoreRequest의 responseType 삭제 및 dto 추가
4. FirebaseAPIManager의 fetch 메서드 파라미터 변경 및 Firestore+Combine 적용
5. 로직 변경하면서 미구현된 부분 임시 삭제

DalioKim committed:
REFACTOR:
1. 디코딩 분리
2. Document(단일) 로직 변경 후 다시 추가
3. 네이밍 수정 및 사용하지 않는 파라미터 제거

DalioKim committed:
REFACTOR:
1. 가독성고려한 네이밍 변경(dto -> query, firestoreTaskPublisher -> firestoreGetTaskPublisher)
2. firestorePostTaskPublisher 네이밍 변경 후 재추가

DalioKim committed:
REFACTOR: PR분리를 위한 삭제 및 접근제한 변경

DalioKim committed:
CHORE: 리베이스 후 누락된 파일들 추가

hanjaesueng commented on GatheRunner/Data/Helpers/CollectionReference+AddQueries.swift:
```diff
@@ -0,0 +1,30 @@
+        }
+        
+        for query in queries {
+            switch query {
+            case .equal(let fieldPath, let condition):
+                whereField(fieldPath, isEqualTo: condition)
+            case .contains(let fieldPath, let condition):
+                whereField(fieldPath, isEqualTo: condition)
+            case .notContains(let fieldPath, let condition):
+                whereField(fieldPath, isEqualTo: condition)
+            case .range(let fieldPath, let start, let end):
+                whereField(fieldPath, isGreaterThanOrEqualTo: start).whereField(fieldPath, isLessThanOrEqualTo: end)
+            }
```
질문
--------------------------
whereField는 어떤 기능인가요?

DalioKim replied:
whereField는 firestore에서 제공하는 기능으로 데이터베이스에 질의요청할때 조건을 추가하는 기능입니다. 저희 앱의 경우에는 달리기기록에 대해서 서버에 요청할때 유저의 아이디와 날짜등을 조건으로 추가하면 해당 조건에 부합하는 데이터만 필터링되어 받을 수 있습니다. 

그리고 CollectionReference+AddQueries 파일은 서버사이드 담당이 없기때문에 화면을 만드는 사람이 쿼리와 테이블에 대해 고려하지 않게끔 테이블명과 조건만 입력되면 쿼리생성을 도와주고요.

hanjaesueng commented on GatheRunner/Data/Helpers/Firestore+Combine.swift:
```diff
@@ -0,0 +1,90 @@
+
+    // MARK: Internal
+
+    func firestoreGetTaskPublisher<D: Decodable>(
+        as _: D.Type) -> AnyPublisher<D, Error>
+    {
+        getDocumentToPublisher
+            .tryMap { try JSONDecoder().decode(D.self, fromJSONObject: $0.data()) }
+            .eraseToAnyPublisher()
+    }
+
+    func firestoreGetTaskPublisher<D: Decodable>(
+        as _: D.Type,
+        querySnapshotMapper: @escaping (QuerySnapshot) -> [D] = QuerySnapshot.defaultMapper()) -> AnyPublisher<[D], Error>
+    {
+        getDocumentsToPublisher
+            .map { querySnapshotMapper($0) }
+            .eraseToAnyPublisher()
+    }
+
```
질문
-----------------------
구체적으로 어떤 용도로 쓰려고 만든 메서드 인가요?

DalioKim replied:
파이어스토어 네트워크 통신의 결과를 퍼블리셔(게시자)로 만들어주는 메서드입니다. 네트워크 요청에 대한 결과(데이터 혹은 에러)등에 대해서 구독자들에게 전달을 해줍니다. 

레포지토리등의 중간계층을 거쳐서 뷰모델에서 퍼블리셔를 구독하면, 통신결과를 받아 볼 수 있고, 그 결과의 상태에 따라 뷰에 필요한 데이터를 가공하고 전달할 수 있습니다.

DalioKim committed:
FIX: 쿼리조건 추가되지 않는 문제 수정

DalioKim committed:
FIX: 다중쿼리 로직 안되는 문제 재귀로직으로 변경해서 수정

DalioKim committed:
REFACTOR: CollectionReference+AddQueries 파일 Query+AddQueries 으로 파일명 수정
FIX: range 컬럼에 단일 조건인 경우, 배열로 추가되어 쿼리적용이 안되는 문제 수정

DalioKim committed:
FEAT: Firestore의 Document를 설정하여 요청하는 메서드 추가

1. FirebaseAPIManager Fetch(withDocument request: FireStoreRequestWithDocument) 메서드 구현
2. FireStoreRequestWithDocument 프로토콜 구현
3. DocumentReference FirestorePostTaskPublisher 메서드, GetDocumentToPublisher 계산속성 구현

REFACTOR: FirebaseAPIManager Fetch 메서드 구분하기 위해, argument 변경

DalioKim commented:
# 주요 변경사항
	- CollectionReference+AddQueries를 Query+AddQueries으로 extension 및 로직 변경 후 쿼리조건이 추가되지 않는 문제 수정
	- Encodable+Queries에서 범위 컬럼에 단일 조건인 경우, 배열로 추가되어 쿼리적용이 안되는 문제 수정
	- fetch withDocument 추가 구현

# 중점적으로 봐줬으면 하는 부분
	- 사용편의성 
	- FireStoreRequest(쿼리와 문서) 구분을 위한 네이밍
	- Query+AddQueries, Encodable+Queries 로직 가독성
	- 예외처리해야할 에러

DalioKim commented:
# Document 설정 사용 예시
        1. 기존방식과 똑같이 콘솔에 테이블을 추가합니다.
        
        2. FireStoreRequestWithDocument을 준수하여 Request를 구현합니다.
       
```
import Foundation

struct WeatherInfoRequest: FireStoreRequestWithDocument {
    var collectionName: String {
        "WeatherInfo"
    }
    
    var documentName: String {
        "Today"
    }
}

struct WeatherInfoResponse: Decodable {
    let content: String?
    let outdoorGrade: String?
    
    enum Keys: String, CodingKey {
        case content = "Content"
        case outdoorGrade = "Outdoor_Grade"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        content = try container.decodeIfPresent(String.self, forKey: .content) ?? ""
        outdoorGrade = try container.decodeIfPresent(String.self, forKey: .outdoorGrade) ?? "0"
    }
}
```

        3. 임의 APIs파일을 통해 정의하고, 호출합니다.

```
import Combine

class APIs {
    static var cancelBag = Set<AnyCancellable>()
    
    static func testFetchWeather() -> AnyPublisher<WeatherInfoResponse, Error> {
        FirebaseAPIManager.shared.fetch(withDocument: WeatherInfoRequest(), as: WeatherInfoResponse.self)
    }
}
```

DalioKim commented:
# 리팩토링 해야하는 목록

- [ ] JSONDecoder 
- [ ] Encodable+Queries Mirror 부분
- [x] FireStoreRequest

DalioKim committed:
REFACTOR: 사용하지 않는 코드 삭제 및 역할 분리

1. FireStoreRequestWithQuery 프로토콜 추가로 Query와 Document 요청 분리
2. QueryOption의 사용하지 않는 case 삭제

DalioKim commented:
# 주요 변경사항
	- FireStoreRequestWithQuery 프로토콜 정의(쿼리로 요청하는 경우 해당 프로토콜을 준수해야하는 방식으로 변경)
	- QueryOption에서 사용하지 않는 case 삭제
	
추가적인 피드백은 firebaseAuth에서 다시 해주시면 감사하겠습니다.
